subprojects/build-cache-http/src/integTest/groovy/org/gradle/caching/NextGenBuildCacheHttpIntegrationTest.groovy

```diff
@@ -51,20 +60,96 @@ class NextGenBuildCacheHttpIntegrationTest extends AbstractIntegrationSpec imple
         """
 
         when:
-        runWithCacheNG "compileJava"
+        runWithBuildCacheNG "compileJava"
         then:
         executedAndNotSkipped ":compileJava"
 
         def alternativeBuildCache = new TestBuildCache(temporaryFolder.file("cache-dir-2").deleteDir().createDir())
         settingsFile << alternativeBuildCache.localCacheConfiguration()
 
         when:
-        runWithCacheNG "clean", "compileJava"
+        runWithBuildCacheNG "clean", "compileJava"
         then:
         skipped ":compileJava"
     }
 
-    private runWithCacheNG(String... tasks) {
+    def "should use different hashes for same artifacts than production build cache"() {
+        def filter = new HashCollectingFilter()
+        httpBuildCacheServer.customHandler.addFilter(new FilterHolder(filter), "/*", EnumSet.of(DispatcherType.REQUEST))
+        buildFile << """
+            plugins {
+                id("base")
+            }
+
+            @CacheableTask
+            abstract class BuildCacheTask extends DefaultTask {
+                @Input
+                abstract Property<String> getInput();
+                @OutputFile
+                abstract RegularFileProperty getOutput();
+
+                @TaskAction
+                void run() {
+                    def file = output.asFile.get()
+                    file.text = input.get()
+                }
+            }
+
+            tasks.register("testBuildCache", BuildCacheTask) {
+                input.set("Hello world")
+                output.set(file("build/build-cache/output.txt"))
+            }
+        """
+
+
+        when:
+        runWithBuildCache "testBuildCache"
+
+        then:
+        executedAndNotSkipped ":testBuildCache"
+        def productionHashes = filter.getAndClearArtifactHashes()
+        !productionHashes.empty
+
+        when:
+        runWithBuildCacheNG "clean", "testBuildCache"
+
+        then:
+        executedAndNotSkipped ":testBuildCache"
+        def ngHashes = filter.getAndClearArtifactHashes()
+        !ngHashes.empty
+        ngHashes.intersect(productionHashes).empty
+    }
+
+    private runWithBuildCache(String... tasks) {
+        withBuildCache().run(tasks)
+    }
+
+    private runWithBuildCacheNG(String... tasks) {
         withBuildCache().run("-D${NextGenBuildCacheControllerFactory.NEXT_GEN_CACHE_SYSTEM_PROPERTY}=true", *tasks)
     }
+
+    private class HashCollectingFilter implements Filter {
+        private final Set<String> artifactHashes = new LinkedHashSet<>()
```

Comment:
:x: I think this can be simplified if you look in the `cacheDir` of the server.

Reply:
Simplified the test and remove filter